services/spar/src/Spar/Scim/User.hs

```diff
@@ -260,42 +295,63 @@ validateScimUser' midp richInfoLimit user = do
             }
       pure richInfo
 
--- | Given an 'externalId' and an 'IdP', construct a 'ST.ValidExternalId'.
+-- | Given an 'externalId' and an 'IdP', construct a 'AuthId'.
 --
 -- This is needed primarily in 'validateScimUser', but also in 'updateValidScimUser' to
 -- recover the 'SAML.UserRef' of the scim user before the update from the database.
-mkValidExternalId ::
+mkAuth ::
   forall m.
   (MonadError Scim.ScimError m) =>
+  TeamId ->
   Maybe IdP ->
   Maybe Text ->
-  m ST.ValidExternalId
-mkValidExternalId _ Nothing = do
+  [ScimEmail.Email] ->
+  m AuthId
+mkAuth _ _ Nothing _ = do
   throwError $
     Scim.badRequest
       Scim.InvalidValue
-      (Just "externalId is required for SAML users")
-mkValidExternalId Nothing (Just extid) = do
+      (Just "externalId is required for SCIM users")
+mkAuth tid Nothing (Just extid) emails = do
   let err =
         Scim.badRequest
           Scim.InvalidValue
-          (Just "externalId must be a valid email address or (if there is a SAML IdP) a valid SAML NameID")
-  maybe (throwError err) (pure . ST.EmailOnly) $ parseEmail extid
-mkValidExternalId (Just idp) (Just extid) = do
-  let issuer = idp ^. SAML.idpMetadata . SAML.edIssuer
-  subject <- validateSubject extid
-  let uref = SAML.UserRef issuer subject
+          (Just "User must have an email address or the externalId must be a valid email address")
+
+  mbEmailFromFields <- primaryEmail emails
+
+  emailWithSource <- case (mbEmailFromFields, parseEmail extid) of
+    (Just emailFromFields, _) ->
+      pure $ EmailWithSource emailFromFields EmailFromEmailField
+    (Nothing, Just emailFromExt) ->
+      pure $ EmailWithSource emailFromExt EmailFromExternalIdField
+    (Nothing, Nothing) ->
+      throwError err
+
+  pure $ AuthSCIM (ScimDetails (ExternalId tid extid) emailWithSource)
+mkAuth tid (Just idp) (Just extid) _ = do
+  uref <- createUserRef idp extid
   pure $ case parseEmail extid of
-    Just email -> ST.EmailAndUref email uref
-    Nothing -> ST.UrefOnly uref
+    Just email -> AuthBoth tid uref (Just (EmailWithSource email EmailFromExternalIdField))
+    Nothing -> AuthSAML uref -- TODO: also handle external email here!
```

Comment:
What is TODO here?

Reply:
24e092f

Reply:
(Not sure if that helped?  Please ask again if not.)

Reply:
Thanks, that helped. Interestingly the impelmentation: `AuthBoth` is guaranteed to have a `EmailWithSource` value, even though we desinged the constructor `AuthBoth TeamId SAML.UserRef (Maybe EmailWithSource)`. I think this makes sense, since having an uref and email is exactly the defintion of `AuthBoth`.